### PR TITLE
Finish deleting the le32/le64 targets

### PR DESCRIPTION
**Pull request by AaronBallman**
This is a revert of ef5e7f90ea4d5063ce68b952c5de473e610afc02 which was a temporary partial revert of 77ac823fd285973cfb3517932c09d82e6a32f46d. The le32 and le64 targets are no longer necessary to retain, so this removes them entirely.

### CODE REVIEW

File: clang/lib/Basic/Targets/OSTargets.h [hunk 1]
```diff
@@ -842,8 +842,8 @@ class LLVM_LIBRARY_VISIBILITY NaClTargetInfo : public OSTargetInfo<Target> {
     } else if (Triple.getArch() == llvm::Triple::mipsel) {
       // Handled on mips' setDataLayout.
     } else {
-      assert(Triple.getArch() == llvm::Triple::le32);
-      this->resetDataLayout("e-p:32:32-i64:64");
+      assert(Triple.getArch() == llvm::Triple::mipsel);
```
Comment (dschuff) [inline clang/lib/Basic/Targets/OSTargets.h:845]: this assert will never pass, since the previous `else if` condition is the same.

Comment (AaronBallman) [reply clang/lib/Basic/Targets/OSTargets.h:845]: Great catch!

**Comment by AaronBallman** [comment]
Precommit CI failure on Windows appears to be unrelated:
```
_bk;t=1720718143606******************** TEST 'lld :: ELF/basic-sparcv9.s' FAILED ********************
_bk;t=1720718143606Exit Code: 1
_bk;t=1720718143606
_bk;t=1720718143606Command Output (stdout):
_bk;t=1720718143606--
_bk;t=1720718143606# RUN: at line 2
_bk;t=1720718143606c:\ws\src\build\bin\llvm-mc.exe -filetype=obj -triple=sparc64-unknown-openbsd C:\ws\src\lld\test\ELF\basic-sparcv9.s -o C:\ws\src\build\tools\lld\test\ELF\Output\basic-sparcv9.s.tmp
_bk;t=1720718143606# executed command: 'c:\ws\src\build\bin\llvm-mc.exe' -filetype=obj -triple=sparc64-unknown-openbsd 'C:\ws\src\lld\test\ELF\basic-sparcv9.s' -o 'C:\ws\src\build\tools\lld\test\ELF\Output\basic-sparcv9.s.tmp'
_bk;t=1720718143606# note: command had no output on stdout or stderr
_bk;t=1720718143606# RUN: at line 3
_bk;t=1720718143606c:\ws\src\build\bin\ld.lld.exe C:\ws\src\build\tools\lld\test\ELF\Output\basic-sparcv9.s.tmp -o C:\ws\src\build\tools\lld\test\ELF\Output\basic-sparcv9.s.tmp2
_bk;t=1720718143606# executed command: 'c:\ws\src\build\bin\ld.lld.exe' 'C:\ws\src\build\tools\lld\test\ELF\Output\basic-sparcv9.s.tmp' -o 'C:\ws\src\build\tools\lld\test\ELF\Output\basic-sparcv9.s.tmp2'
_bk;t=1720718143606# note: command had no output on stdout or stderr
_bk;t=1720718143606# RUN: at line 4
_bk;t=1720718143606c:\ws\src\build\bin\llvm-readobj.exe --file-headers --sections -l --symbols C:\ws\src\build\tools\lld\test\ELF\Output\basic-sparcv9.s.tmp2    | c:\ws\src\build\bin\filecheck.exe C:\ws\src\lld\test\ELF\basic-sparcv9.s
_bk;t=1720718143606# executed command: 'c:\ws\src\build\bin\llvm-readobj.exe' --file-headers --sections -l --symbols 'C:\ws\src\build\tools\lld\test\ELF\Output\basic-sparcv9.s.tmp2'
_bk;t=1720718143606# note: command had no output on stdout or stderr
_bk;t=1720718143606# executed command: 'c:\ws\src\build\bin\filecheck.exe' 'C:\ws\src\lld\test\ELF\basic-sparcv9.s'
_bk;t=1720718143606# .---command stderr------------
_bk;t=1720718143606# | C:\ws\src\lld\test\ELF\basic-sparcv9.s:20:15: error: CHECK-NEXT: expected string not found in input
_bk;t=1720718143606# | # CHECK-NEXT: OS/ABI: SystemV (0x0)
_bk;t=1720718143606# |               ^
_bk;t=1720718143606# | <stdin>:12:16: note: scanning from here
_bk;t=1720718143606# |  FileVersion: 1
_bk;t=1720718143606# |                ^
_bk;t=1720718143606# | <stdin>:13:2: note: possible intended match here
_bk;t=1720718143606# |  OS/ABI: OpenBSD (0xC)
_bk;t=1720718143606# |  ^
_bk;t=1720718143606# | 
_bk;t=1720718143606# | Input file: <stdin>
_bk;t=1720718143606# | Check file: C:\ws\src\lld\test\ELF\basic-sparcv9.s
_bk;t=1720718143606# | 
_bk;t=1720718143606# | -dump-input=help explains the following input dump.
_bk;t=1720718143606# | 
_bk;t=1720718143606# | Input was:
_bk;t=1720718143606# | <<<<<<
_bk;t=1720718143606# |            .
_bk;t=1720718143606# |            .
_bk;t=1720718143606# |            .
_bk;t=1720718143606# |            7: ElfHeader { 
_bk;t=1720718143606# |            8:  Ident { 
_bk;t=1720718143606# |            9:  Magic: (7F 45 4C 46) 
_bk;t=1720718143606# |           10:  Class: 64-bit (0x2) 
_bk;t=1720718143606# |           11:  DataEncoding: BigEndian (0x2) 
_bk;t=1720718143606# |           12:  FileVersion: 1 
_bk;t=1720718143606# | next:20'0                    X error: no match found
_bk;t=1720718143606# |           13:  OS/ABI: OpenBSD (0xC) 
_bk;t=1720718143606# | next:20'0     ~~~~~~~~~~~~~~~~~~~~~~~
_bk;t=1720718143606# | next:20'1      ?                      possible intended match
_bk;t=1720718143606# |           14:  ABIVersion: 0 
_bk;t=1720718143606# | next:20'0     ~~~~~~~~~~~~~~~
_bk;t=1720718143606# |           15:  Unused: (00 00 00 00 00 00 00) 
_bk;t=1720718143606# | next:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
_bk;t=1720718143606# |           16:  } 
_bk;t=1720718143606# | next:20'0     ~~~
_bk;t=1720718143606# |           17:  Type: Executable (0x2) 
_bk;t=1720718143606# | next:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~
_bk;t=1720718143606# |           18:  Machine: EM_SPARCV9 (0x2B) 
_bk;t=1720718143606# | next:20'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
_bk;t=1720718143606# |            .
_bk;t=1720718143606# |            .
_bk;t=1720718143606# |            .
_bk;t=1720718143606# | >>>>>>
_bk;t=1720718143606# `-----------------------------
_bk;t=1720718143606# error: command failed with exit status: 1
_bk;t=1720718143606
_bk;t=1720718143606--
_bk;t=1720718143606
_bk;t=1720718143606********************
```

**Review by MaskRay — APPROVED** [review]
LGTM! That partial Le32/Leb64 restore was to give Halide some time.

**Comment by steven-johnson** [comment]
Unfortunately, this utterly breaks Halide, which was in fact relying on both of these targets.

**Comment by steven-johnson** [comment]
> LGTM! That partial Le32/Leb64 restore was to give Halide some time.

I understand that you want these gone, but the fact of the matter is that Halide isn't ready for this change, and so the next integrate of LLVM into google3 will break the world. We urgently need these features (or something like them).

**Comment by efriedma-quic** [comment]
It looks like the removal missed a few bits: PNaClABIBuiltinVaList and clang/lib/CodeGen/Targets/PNaCl.cpp weren't removed.

> Unfortunately, this utterly breaks Halide, which was in fact relying on both of these targets.

We can temporarily revert this, but we aren't going to leave the orphaned remains of PNaCl in-tree forever.  If you need a "halide" target long-term, please write up a proposal on Discourse describing exactly what bits you need, why you need them, relevant test coverage in-tree and out-of-tree, and a proposed code owner for those bits.

**Comment by MaskRay** [comment]
It seems that Halide (https://github.com/halide/Halide/pull/5934), an important user, still needs generic 32-bit 64-bit Clang targets.
The removed le32/le64 targets fulfilled this need.
(This change broke my internal users. Sorry that I did not realize that Halide#5934 was closed without an actual fix @abadams).

@AaronBallman Are you comfortable with bring back le32/le64 temporarily?
The few occurrences in `bitfield-*` don't serve useful testing purposes, and they can be removed if they feel like eyesore.

**Comment by abadams** [comment]
Our actual need is for a stable and user-friendly way to write fragments of mostly-target-agnostic llvm IR.

We concat these with other fragments and give it an actual target triple later. This lets us assemble a target-specific Halide runtime dynamically by mixing and matching some number of these fragments. Compiling these fragments for all possible target triples we support instead of a generic one would multiply our code size and compile time by the (large) number of possible target triples we can emit. Not using fragments at all and compiling a runtime in a more conventional way would be a combinatorial amount of code, because there are a lot of combinations. Writing generic .ll directly is possible, and we do a little of that for very low-level things, but writing in C++ and compiling via clang to le64/32 is more stable and much more pleasant due to the abstractions that C++ provides. We just have to be careful to not do things like have structs that would have different layouts on different platforms. 

If there's a better way to do all this, let us know. Does no other project have a need for "generic" llvm IR without any target-specific annotations on it yet?

**Comment by efriedma-quic** [comment]
There is no such thing as "generic" LLVM IR, in general: so many target-specific aspects of C have to be directly encoded into the IR, so there's no way to actually make it generic.  In this respect, the "le32/le64" targets aren't really any more generic than any other target; they encode their own specific assumptions, and they're not ABI-compatible with code generated for any other target.  The closest you can get to being "generic" is picking a target that's intended to be compiled to other targets, like SPIRV/wasm/etc.

So the question is, why specifically use the PNaCl targets, instead of just using the existing Arm or wasm targets?

**Comment by abadams** [comment]
We were advised to use the pnacl targets at the time the code was written.  We're trying wasm now.

The IR can be generic if you use a subset of C that excludes all the things that vary across the targets you care about. For cases we really can't avoid it (e.g. win32 calling convention stuff for direct3d), we compile that one fragment with a few different target triples.

**Comment by steven-johnson** [comment]
> We're trying wasm now.

This looked promising in our GitHub tests, but for reasons that aren't clear (yet), ~everything compiled this way inside google3 resulted in linker errors (missing symbols).

**Comment by efriedma-quic** [comment]
> The IR can be generic if you use a subset of C that excludes all the things that vary across the targets you care about

You can sort of make this work, I guess.  We don't document any such subset, though, and some of the issues aren't obvious at first glance.

**Comment by steven-johnson** [comment]
After some initial experiments, it looks unlikely that a quick fix is possible on our side. May I suggest a revert with a (say) 30-day expiration date? That should give us enough time to come up with a fix on our side (which will actually be a drop-other-work-and-fix-it approach this time).

**Comment by AaronBallman** [comment]
> After some initial experiments, it looks unlikely that a quick fix is possible on our side. May I suggest a revert with a (say) 30-day expiration date? That should give us enough time to come up with a fix on our side (which will actually be a drop-other-work-and-fix-it approach this time).

Thank you for the investigations -- thirty days seems extremely reasonable to me. If you run into issues while fixing it on your end and need more than thirty days, that's also fine (we can figure out how much longer to wait at that point).

I'll get the revert done shortly.

**Comment by AaronBallman** [comment]
d3f8105c65046173e20c4c59394b4a7f1bbe7627 restores the targets, thank you all for the discussion!

**Comment by efriedma-quic** [comment]
> I was under the impression those bits were still needed for Chromium. CC @dschuff

Oh, clang/lib/CodeGen/Targets/PNaCl.cpp is still live for mipsel-nacl target triple.  Which we also have zero tests for, but I guess that's a separate issue.

**Comment by steven-johnson** [comment]
Note that Halide has removed its dependency on the le32/le64 targets as of https://github.com/halide/Halide/pull/8344.

**Comment by AaronBallman** [comment]
> Note that Halide has removed its dependency on the le32/le64 targets as of [halide/Halide#8344](https://github.com/halide/Halide/pull/8344).

Thank you for letting us know!

Does anyone know of any reasons we should not revert the revert (so le32/le64 is removed in Clang 20)?

**Pull request merged**